src/lib/pipeline/BlockingIssueUnit.cc

```diff
@@ -0,0 +1,270 @@
+
+    if (uop->isLoad()) {
+      lsq_.addLoad(uop);
+    }
+    if (uop->isStoreAddress()) {
+      lsq_.addStore(uop);
+    }
+
+    issuePorts_[port].getTailSlots()[0] = std::move(uop);
+    portAllocator_.issued(port);
+
+    issueQueue_.pop_front();
+    issued++;
+  }
+
+  // Record reasoning for zero issuing this cycle
+  if (issued == 0) {
+    if (input_.isStalled())
+      backendStalls_++;
+    else
+      frontendStalls_++;
+  }
+}
+
+void BlockingIssueUnit::forwardOperands(const span<Register>& registers,
+                                        const span<RegisterValue>& values) {
+  assert(registers.size() == values.size() &&
+         "Mismatched register and value vector sizes");
+
+  for (size_t i = 0; i < registers.size(); i++) {
+    const auto& reg = registers[i];
+    // If the forwarded register is one depended on, supply value to uop at
+    // front of issue queue
+    if (dependent_ && reg == dependency_.first) {
+      auto& uop = issueQueue_.front();
+      uop->supplyOperand(dependency_.second, values[i]);
+      assert(uop->getSequenceId() == issueQueue_.front()->getSequenceId() &&
+             "Tried to early issue uop not at front of queue");
+      // If the uop is now ready to execute, identify whether it can be issued
+      // to avoid pipeline bubbles
+      if (issueQueue_.front()->canExecute()) {
+        bool ready = true;
+
+        // Allocate issue port to uop
+        uint16_t port = portAllocator_.allocate(uop->getSupportedPorts());
+
+        // Query whether port is free
+        if (issuePorts_[port].getTailSlots()[0] == nullptr) {
+          auto& destinationRegisters = uop->getDestinationRegisters();
+          // Query whether the destination registers are ready
+          for (const auto& reg : destinationRegisters) {
+            if (!scoreboard_[reg.type][reg.tag].first) {
+              ready = false;
+              break;
+            }
+          }
+        } else {
+          ready = false;
+        }
+
+        // Issue the uop if all required resources are available, else
+        // deallocate the port
+        if (ready) {
+          auto& destinationRegisters = uop->getDestinationRegisters();
+          for (const auto& reg : destinationRegisters) {
+            scoreboard_[reg.type][reg.tag] = {false, uop->getSequenceId()};
+          }
+
+          recordIssue_(uop);
+
+          if (uop->isLoad()) {
+            lsq_.addLoad(uop);
+          }
+          if (uop->isStoreAddress()) {
+            lsq_.addStore(uop);
+          }
+
+          issuePorts_[port].getTailSlots()[0] = std::move(uop);
+          portAllocator_.issued(port);
+
+          issueQueue_.pop_front();
+          input_.stall(false);
+        } else {
+          portAllocator_.deallocate(port);
+        }
+      }
+
+      dependent_ = false;
+      dependency_ = {};
+      break;
+    }
+  }
+}
+
+void BlockingIssueUnit::setRegisterReady(Register reg) {
+  scoreboard_[reg.type][reg.tag] = {true, -1};
```

Comment:
Again, using -1 in a uint64 could be problematic for very large simulation runs where this ID (i.e. uint64-max) is in use